### DISCUSSION TITLE
Barracuda_rooted_libre-11.1842.0 (Last version with Cortana support)

### DISCUSSION THREAD

**Discussion by coggy9** [coggy9/HKHacking#7] [General]
[This](https://github.com/coggy9/HKHacking/releases/tag/StockRoot) is a system image with an adb shell over the network enabled, and updates disabled by blocking the domains through the hosts file. The boot sound has been replaced to indicate that this is a custom, unsafe, image, and should **only** be used for development/research. And it's loud on the first boot, sorry.

For some reason, the setup/wifi AP creation process will fail to start sometimes, just hold the mic button for five seconds if this happens.

Use the tools released earlier to flash this file.

<hr /><em>This discussion was created from the release <a href='https://github.com/coggy9/HKHacking/releases/tag/StockRoot'>Barracuda_rooted_libre-11.1842.0</a>.</em>

**Comment by Stedmister** [comment]
this is really neat guys, though I'm confused, been watching this process unfikd great work.

**Reply by coggy9** [reply]
This isn't for end users yet. This is just for developers right now, so they can poke around a mostly stock system.

**Comment by rbeesley** [comment]
I've been digging through this image this weekend, trying to understand it a bit more. Very interesting. So much seems based on a Chromecast as the base and Libre. My guess is that the system image started as a Chromecast, then it was made into a Libre, then HK took that to make the Podium, and then took that and added Cortana to make the Invoke.

How much did you change to make this image? There were a couple of references I saw which commented out "disabled" for adb and dropbear which I think is how adb was enabled. The version string Libre_rooted, is that how it came? Also seems that most of it is from around October 2018, so it doesn't seem as though it is the most recent firmware image, but it seems to offer a lot.

~~The animations for the LEDs are some binary format, unfortunately it looks like running them is part of the Cortana binary. It would have been nice if those were driven by an external process. I think they are probably driven via gpio, but I hadn't confirmed that yet.~~

Animations might be handled through mcu-interface. You can see the commands sent when monitoring logcat over adb.

```
I/mcu-interface( 1113): received cmd:  4  9  1  0  0  0
I/mcu-interface( 1113): MCU key event Volume down received!
I/mcu-interface( 1113): Receive volume change notify event: 24!
I/mcu-interface( 1113): sending cmd:  3 18 ae b6  0  0
```

If that's the case, and controlling the UI as well as reading input such as taps and volume control are handled via the mcu-interface, then we'd just have to map that.

The service port seems to be initialized during boot, but I haven't figured out how it is supposed to be connected. Leveraging it as a console port would be nice. Dropbear seems to be started, but I couldn't figure out a way to actually log in. I think we'd need to put an ssh public key in the home, but that's a ro partition.

I was able to tar up the entire mounted image for offline browsing. The /run mount is rw and has plenty of room to archive the entire system, then I was able to pull that over adb.

To fix the echo problem in the shell, you can run reset which seems to fix things. Running the bash shell is probably smart anyway and I've also been changing the prompt to see where I am when I'm perusing.

There's a lot more to look at in this image than in 82_IMAGE. Probably a lot I haven't seen yet too... I haven't tried to load all the mtdblock partitions as I did with 82_IMAGE, so there's that. It seems to do some interesting things generating keys, and I saw where that is done on first boot, but I haven't quite figured out how it sets up the AP on first boot.

Lots to take in, but it seems like there's a lot of opportunity still. Running as a BT dumb speaker is greatly underutilizing the hardware.

We should also allow for the update to download, just not install, and see if we can write an image file we can flash to apply the BT speaker update manually before the servers are shut down in June.

**Reply by coggy9** [reply]
IIRC, these are all the changes I made:

- Uncommented `write /sys/class/android_usb/android0/f_acm/instances "1"` in init.rc under the `# Enable ADB and ACM` section in `init.rc`
- Commented out the "disabled" line for the adbd service and `start adbd` in `init.rc`
- Commented out the retail firewall rules in `/usr/sbin/firewall.sh`, and moved the debug rules out of it's if statement
- Replaced a few of the prerecorded Cortana lines, and the boot sound
- Added the "_rooted" string to the version wherever I found it as another way to indicate that this isn't a stock image. 
- Added the OTA server domains to `/etc/hosts`

I'm 99% sure that this is the latest Cortana enabled firmware. I'll make another image with OTA servers unblocked so we can attempt to capture the current update.

**Comment by coggy9** [comment]
Here's an image with the OTA block removed, but I don't think we're getting the update. OTA checking seems to fail for some unknown reason, and even normal customers are left without the update, so it's probably something on Harman's end. 

https://1drv.ms/u/s!AlMClBTJnoF0gvARltrIcyAsjEXF9w?e=JFdUQ2

**Reply by coggy9** [reply]
![image](https://user-images.githubusercontent.com/352207/115292724-f619b180-a11b-11eb-86fe-1f91715f47b3.png)
I may have spoke too soon.

**Comment by rbeesley** [comment]
I looked at this some more this weekend. From the installed rooted firmware I was able to extract the image binaries from the `/dev/mtd devices`. From that I was able to use `binwalk` to extract the filesystems from `07-bootimgs_B`, `08-bsl`, `09-bootimgs`, `10-rootfs`, and `15-mv_nand`. 7, 9, and 15 seem to have `cpio` containers within them.  `08-bsl` seems like the base OS volume since most of the files contained seem like they'd produce a working bootable image with directories used as mount points. Bootstrap Loader? It isn't just overlaid as it doesn't look like it is directly referenced, but at least one file from this block device is (copied) in `/lsync`. It does have /etc/init.d/rcS which references mounting `/lsync` from `/dev/mtdblock11`, so I don't yet know what to think of this. `10-rootfs` seems like it is the root filesystem as the mapped name might suggest. I'm not sure what `15-mv_nand` is either.  There is one executable which doesn't seem to have any meaningful strings to me. It has an ELF header, so it is an executable, but I'm not sure what it is: `1A20060: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), statically linked, missing section headers`.

If we can block rebooting after the firmware downloads, I would be able to go back and extract the firmware from the block devices again so we'd have something to compare against. If we can get to that point, then we would likely be able to rebuild the image and root it as well. This way we don't have to decode the encrypted image, only the resulting image.

**Reply by coggy9** [reply]
The problem with trying to apply the update on a rooted/modified firmware, is that the update package is a delta update, and is expected to be applied over a unmodified system, and cannot be applied if anything is changed. 

We need a way to let the update apply on stock, and extract the images using the bootloader somehow, but the bootloader has its own set of issues when trying to read the NAND out to the terminal.

**Reply by rbeesley** [reply]
I think that was what I was addressing. Or trying to anyway.

The Redbend update process (rb_ua I think?) is what would write to the nand. Then if we block rebooting at that point, we'd be able to pull the modified images directly from the mtd block devices. Essentially the update would apply as it is supposed to, but we'd have the resolved composite of the modified image... Or are you suggesting the flashing actually takes place on the reboot? Even if that's the case it would need to write the update to some non-volatile region which would let us see the state before it updates allowing offline flashing to the final build at that point.

I have some experience with Linux and Android firmware, but there's a lot that I'm learning on the fly on this project, so I'm sure there's some things I could be missing too.

**Comment by coggy9** [comment]
Still waiting for Harman's V12 flashing package, since trying to properly compile anything for V11 is nearly impossible, at least with my very limited experience, but I've been trying to get music to play just using the binaries included in the stock firmware. 

There is a binary called `music-test` that will play anything I throw at it, including MP3 streams, but the issue is that it changes the volume every few seconds, usually to the max.  If whatever function is changing the volume could be patched out, streaming music functionality could be partially restored. 

[Here](https://b.catgirlsare.sexy/kd7rQFCR.) is the `music-test` binary if anyone skilled in ARM disassembly wants to take a look without downloading the whole image.

There is also the `gst-play-1.0` binary, but it doesn't route the audio through the speaker, like `music-test` does, and `gst-device-monitor-1.0` [can't start](https://i.imgur.com/jrkYblN.png), so it's probably a deeper issue with the standard GStreamer binaries included, excluding `music-test` since it appears to use a version of GStreamer built into the binary to play music, going by a string dump.

**Comment by coggy9** [comment]
[Looks like Microsoft has taken down the "legacy/V1" Cortana servers.](https://twitter.com/coggy9/status/1417754527177056256)

Invokes stuck on V11 may not be able to go through setup anymore, but setup should be skippable with this modified FW. I'll see about getting a usable non developer/rooted image out that boots into an already setup state, so there is _something_ for when someone finds an Invoke that has never been used.

**Reply by Stedmister** [reply]
Thanks for all your work

**Reply by rbeesley** [reply]
I've had to step back from this for a bit... might have some time to devote again. TBD.
